src/mono/mono/metadata/object.c

```diff
@@ -8124,6 +8124,25 @@ mono_runtime_get_managed_cmd_line (void)
 	return cmd_line ? g_string_free (cmd_line, FALSE) : NULL;
 }
 
+void
+mono_runtime_run_startup_hooks (void)
+{
+	if (mono_runtime_get_no_exec ())
+		return;
+
+	MonoClass *klass = mono_class_try_load_from_name (mono_defaults.corlib, "System", "StartupHookProvider");
```

Comment:
Any reason why we won't call the same startup hook logic as CoreCLR and remove the call in ep_rt_mono_init_finish on Mono? If there might be a scenario in the future where the managed parts gets included without EventPipe on Mono (currenlty not the case), then we would need to change this since ep_finish_init would then just be a stub implementation.

Reply:
> Any reason why we won't call the same startup hook logic as CoreCLR and remove the call in ep_rt_mono_init_finish on Mono? If there might be a scenario in the future where the managed parts gets included without EventPipe on Mono (currenlty not the case), then we would need to change this since ep_finish_init would then just be a stub implementation., but the user still expect to get these events delivered over alternative provider. We could of course do that when needed, or we could just do it now, since we get the support that previously was missing and why init of runtime events ended up in ep_rt_mono_init_finish. Anything with the current startup hook implementation that would prevent us from getting the same behavior as we currently have, but done from startup hook instead of ep_rt_mono_init_finish?

There are two separate issues:

1. Support EventSource on Mono without EventPipe.
2. Use the same single `ManagedStartup` call as CoreCLR.

I think it's probably reasonable to do something about 1 - we can move the invoke of RuntimeEventSource.Initialize out of EP and into `mini_init`, for example.

For 2, the issue is that `ManagedStartup` cannot be trimmed - there is no way to tell the trimmer "remove a method if 2 feature switches are both disabled".  So we would always be paying for a call to managed even if event source and startup hooks are both disabled (for example on single-threaded WASM).

So if you think it makes sense to start decoupling EventSource and EventPipe in Mono now, we can do that.  But I don't think it makes sense to do 2.  Of course if we had some other managed initialization that had to be done explicitly on all platforms, then we could roll all that logic into one function.

Reply:
OK, understood, if we still won't end up on the same solution as CoreCLR, then I believe we can keep things as is for now, and move the call to init when/if we ever have a need to separate init of runtime event source from EventPipe.